File: inc.graphql-resources/backend/nodejs/utils/i18n.js
const path = require('path');
const fs = require('fs');

// Cache for loaded language files to avoid reading from disk repeatedly
const langCache = {};

// Helper function for sanitizing HTML output
const esc = (str) => {
    if (str === null || typeof str === 'undefined') return '';
    return String(str).replace(/[&<>"']/g, (match) => {
        return { '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[match];
    });
};

const loadLanguage = (langId) => {
    if (langCache[langId]) {
        return langCache[langId];
    }

    const langFilePath = path.join(__dirname, '../public/langs/i18n', `${langId}.json`);

    try {
        const langData = fs.readFileSync(langFilePath, 'utf8');
        const translations = JSON.parse(langData);
        langCache[langId] = translations;
        return translations;
    } catch (error) {
        console.error(`Failed to load language file for ${langId}:`, error.message);

        // Fallback to 'en' (English) if the requested language file is not found
        if (langId !== 'en') {
            return loadLanguage('en'); // Recursive fallback to English
        }

        return {}; // Return empty object if even the default fails
    }
};

const getTranslator = (req) => {
    const langId = req.headers['x-language-id'] || 'en'; // Default to English

    const translations = loadLanguage(langId);

    return (key, ...args) => {
        let translatedText = translations[key] || key; // Fallback to key if translation missing

        // Replace placeholders {0}, {1}, etc.
        args.forEach((arg, index) => {
            translatedText = translatedText.replace(new RegExp(`\\{${index}\\}`, 'g'), arg);
        });

        return translatedText;
    };
};

module.exports = { getTranslator, esc };

File: inc.graphql-resources/backend/python/utils/i18n.py
import json
from pathlib import Path
from typing import Callable, Dict, Any

from fastapi import Request

# Cache to store loaded translation data
_translations: Dict[str, Dict[str, str]] = {}
_base_dir = Path(__file__).resolve().parent.parent / "static" / "langs" / "i18n"


def _load_translation(lang: str) -> Dict[str, str]:
    """Loads the translation JSON file for a specific language."""
    if lang in _translations:
        return _translations[lang]

    lang_file = _base_dir / f"{lang}.json"
    if not lang_file.exists():
        # Fallback to English if the file is not found
        lang = "en"
        lang_file = _base_dir / "en.json"

    try:
        with open(lang_file, "r", encoding="utf-8") as f:
            translation_data = json.load(f)
            _translations[lang] = translation_data
            return translation_data
    except (IOError, json.JSONDecodeError):
        # If an error occurs, return an empty dictionary
        return {}


async def get_translator(request: Request) -> Callable[..., str]:
    """FastAPI dependency to get the translator function based on the request header."""
    lang = request.headers.get("X-Language-Id", "en")
    translations = _load_translation(lang)

    def translator(key: str, *args: Any) -> str:
        """Translates a key and formats the string if arguments are provided."""
        return translations.get(key, key).format(*args)

    return translator
